validator/client/src/main/java/tech/pegasys/teku/validator/client/KeyManager.java

```diff
@@ -90,20 +100,49 @@ public DeleteKeysResponse deleteValidators(final List<BLSPublicKey> validators)
    *
    * @param keystores strings of keystore files
    * @param passwords strings of passwords
+   * @param slashingProtectionImporter slashing protection importer with loaded context
    * @return a list of 1 status per keystore that was attempted to be imported
    */
-  public List<ImportStatus> importValidators(
-      final List<String> keystores, final List<String> passwords) {
-    throw new NotImplementedException("importValidators not implemented yet");
+  public List<PostKeyResult> importValidators(
+      final List<String> keystores,
+      final List<String> passwords,
+      final Optional<SlashingProtectionImporter> slashingProtectionImporter) {
+    final List<PostKeyResult> importResults = new ArrayList<>();
+    for (int i = 0; i < keystores.size(); i++) {
+      importResults.add(
+          importValidatorFromKeystore(
+              keystores.get(i), passwords.get(i), slashingProtectionImporter));
+    }
+    return importResults;
   }
 
-  /**
-   * Import slashing protection data
-   *
-   * @param slashingProtection slashing protection data in interchange format
-   * @return Error message on failure, otherwise Optional.empty() on success.
-   */
-  public Optional<String> importSlashingProtection(final String slashingProtection) {
-    return Optional.of("Import slashing protection is not yet implemented.");
+  public DataDirLayout getDataDirLayout() {
+    return dataDirLayout;
+  }
+
+  private PostKeyResult importValidatorFromKeystore(
+      final String keystoreString,
+      final String password,
+      final Optional<SlashingProtectionImporter> slashingProtectionImporter) {
+    final KeyStoreData keyStoreData = KeyStoreLoader.loadFromString(keystoreString);
+    try {
+      keyStoreData.validate();
+    } catch (KeyStoreValidationException ex) {
+      return PostKeyResult.error(ex.getMessage());
+    }
+    final BLSPublicKey publicKey =
+        BLSPublicKey.fromBytesCompressed(Bytes48.wrap(keyStoreData.getPubkey()));
+    if (validatorLoader.getOwnedValidators().getValidator(publicKey).isPresent()) {
+      return PostKeyResult.duplicate();
+    }
+
+    if (slashingProtectionImporter.isPresent()) {
+      final Optional<String> errorString =
+          slashingProtectionImporter.get().updateSigningRecord(publicKey, LOG::debug);
+      if (errorString.isPresent()) {
+        return PostKeyResult.error(errorString.get());
+      }
+    }
```

Comment:
I think importing the slashing protection here, outside the synchronised block in `ValidatorLoader` introduces some race conditions because we might wind up having two simultaneous requests importing slashing protection for the same validator and they'd conflict.  We might also be reloading validators from a `HUP` signal and creating new slashing protection.  So I think this winds up needing to move into `validatorLoader.loadMutableValdiator` to be protected by the same lock it uses.

Reply:
We can then also remove the check to see if `ownedValidators` already has the key because it's done inside `validatorLoader` anyway.

Reply:
that's a bit cleaner, i like it. It's starting to look like the keymanager also has fairly limited responsibilities.. just a thought at this stage, might want to review at some point.